### ISSUE TITLE
"vscode://my.extension" links not working in hover messages

### ISSUE THREAD

**Issue by rmzetti** [microsoft/vscode#151720]
Issue Type: <b>Bug</b>

In 1.68.0, links of the form "vscode://my.extension?xxx" have stopped working in trusted hover messages.
Example:
The following link executes the copyToClipboard command in my extension:
      [ Copy to clipboard ](vscode://rmzetti.hover-exec?copyToClipboard)
If it is ctrl-clicked in a markdown file in the editor it executes the extension copyToClipboard command.
If the same link is included in a trusted markdown string hover message, the link is now (v 1.68.0) displayed as plain text and is not clickable.
In version 1.67.1 and previous, both work fine.
The same issue is present in the 1.68.0 for Windows and for Linux/deb.
(I have unpublished my extension until I can resolve this)

VS Code version: Code 1.68.0 (4af164ea3a06f701fe3e89a2bcbb421d2026b68f, 2022-06-08T11:44:16.822Z)
OS version: Windows_NT x64 10.0.22621
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 4700U with Radeon Graphics          (8 x 1996)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.41GB (6.53GB free)|
|Process Argv|--folder-uri file:///c%3A/Users/ralph/OneDrive/Documents/GitHub/download/uri-handler-sample --crash-reporter-id 7678baea-f96e-47d7-b502-ce2fe4788442|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (21)</summary>

Extension|Author (truncated)|Version
---|---|---
markdown-preview-github-styles|bie|1.0.1
vscode-pandoc|Chr|0.4.1
quick-html-previewer|dai|1.0.1
vscode-eslint|dba|2.2.2
windows-explorer-context-menu|ele|3.2.1
copilot|Git|1.27.6094
vscode-pull-request-github|Git|0.44.0
vscode-insertdatestring|jsy|2.3.1
language-julia|jul|1.6.17
vscode-markdown-notes|kor|0.0.24
remote-wsl|ms-|0.66.3
vscode-paste-image|mus|1.0.4
hover-exec|rmz|0.7.3
markdown-preview-enhanced|shd|0.6.3
autoimport|ste|1.5.4
markdown-kit|svs|1.2.0
markdown-memo|svs|0.3.18
markdown-links|tch|0.8.0
pdf|tom|1.2.0
vscodeintellicode|Vis|1.2.21
markdown-all-in-one|yzh|3.4.3

(3 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242:30382549
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
vscscmwlcmt:30465135
cppdebug:30492333
vsclangdf:30486550

```

</details>

**Comment by rmzetti** [comment]
Note - the new experimental validations are NOT enabled

**Comment by JeromeCugny** [comment]
I have the same issue; I'm developing an extension with a custom language.
I have a link in markdown with a custom scheme (`wamdb`) to open links upon hover.
The link is not clickable with the latest version 1.68 but is clickable in 1.67.2.
Note: if I change the scheme to something existing like `file` then it is clickable.

Windows 10 x64
experimental validations is also disabled

**Comment by rmzetti** [comment]
@mjbvz - Thanks for your comment.
 
The main concept in my extension is to provide a facility for various actions on code blocks – executing a range of scripts, etc. For basic execution functions, an execute command is provided, specific to particular code blocks. 
 
However there is also a range of related commands customised appropriately for specific code blocks – these are presented via the markdown hover message as links (as per the vscode extension docs). The ability to see options via the hover and execute them (or execute the code block) is fundamental to the extension.
 
This facility has worked well in previous vscode versions, and these markdown command links still work within the editor (which is obviously not as useful as in a hover message). Also, other links continue to work in the markdown hover message.
 
The issue simply is that the markdown hover message (unlike markdown in the editor) has, after the version update, stopped recognising “vscode://my.extension...” as a hyperlink. This looks like an inadvertent side-effect of another mod .

**Comment by mjbvz** [comment]
@rmzetti Yes but why don’t you just use command links? Those are the preferred way of triggering actions in your extension on click

**Comment by rmzetti** [comment]
@mjbvz Thanks again - I appreciate your thoughts.

The `command:` links you suggest refer to extension contributed commands. I have a couple of those which can indeed be used in the markdown hover message - and they still work in the new version in hover messages (as do `file://` links which I also use).

Most of the 'commands' I am making available in the various hover messages, however, are specific to the hover message itself - they are not contributed 'extension commands' and would make no sense in the command palette. The `vscode://my.extension?cmd` (where 'cmd' here is not a contributed command) is intended to support that process, and as I said still does work fine as `[link](vscode://my.extension?cmd)` in the editor. It just does not currently appear as a link in the hover in the latest version, and therefore can no longer be clicked/utilised as intended.

Hopefully that helps explain what is happening. Thanks again for your input.

**Comment by mjbvz** [comment]
You can register commands that don't appear in command palette: just don't list in in the `contributions` section of your `package.json`

For security/complexity reasons, I'm not in favor of adding back in support for `vscode:` commands unless there is a good use case. Based on what you've shared, I don't think there is

**Comment by rmzetti** [comment]
@mjbvz Ok - thanks.. I haven't tried to utilise commands that are not listed as contributed .. that would certainly solve my issue.. I'll give it a try.

Does that mean `vscode://..` links are to be deprecated, because, although the conversation has drifted into the way I've structured my extension, the actual point was that all the links like:

- `command: ...`
- `file://...`
- `https:// ...`
-  `vscode:// ...`

work in the md editor, and have all worked in hover messages until v1.68 when the last one (only) fails.

If there were a real security issue (rather than simply a reluctance to address the bug), I would expect to see `vscode://..` links deprecated - clearly such an issue relates to both situations. If those links are not to be deprecated, I would appreciate your support for a fix - `vscode` is currently both powerful and consistent - both aspects are important.

**Comment by JeromeCugny** [comment]
@mjbvz 
Hi again, as for me I don't even run a custom command behind the link .. It triggers a `vscode.TextDocumentContentProvider` thanks to my custom scheme, as a http:// would open your browser.

I answered here 'cause I thought the issue mentioned was the click on the link itself and not what it does behind.
> the link is now (v 1.68.0) displayed as plain text and is not clickable  

Do you want me to create another issue? For me it is a regression.  

@rmzetti Sorry to over extend on your issue, I don't mean to change the topic at all. I'm just trying to give my intakes, as for me it's the same issue of clickable link (no matter if it's a command behind or not) :)
_______________________________________
Simple example:
1) register an hover provider for a given language
```typescript
vscode.languages.registerHoverProvider(languageID, {
  async provideHover(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken): Promise<vscode.Hover> {
    return new vscode.Hover(new vscode.MarkdownString(`[link](mjbvz:///some/stuff)`))
  }
})
```
2) Open a file, set the language, write something then hover  

Version 1.67.2:
![image](https://user-images.githubusercontent.com/22790678/174554222-2f0ab719-1c79-4036-b4fd-459fe5f4d685.png)
Version 1.68.1:
![image](https://user-images.githubusercontent.com/22790678/174555223-6d2230ac-9f21-4472-bf03-f244272581b1.png)

**Comment by mjbvz** [comment]
@JeromeCugny You can open a new feature request. Custom links have not been intentionally supported (and I don't know why you saw them working previously): https://github.com/microsoft/vscode/blob/57037a722805136ca8b9f736d6a250d2433c43e5/src/vs/base/browser/markdownRenderer.ts#L348

Command Uris still work for that use case though: simply have your command open the custom document

**Comment by JeromeCugny** [comment]
> @JeromeCugny You can open a new feature request. Custom links have not been intentionally supported (and I don't know why you saw them working previously):
> 
> https://github.com/microsoft/vscode/blob/57037a722805136ca8b9f736d6a250d2433c43e5/src/vs/base/browser/markdownRenderer.ts#L348
> 
> Command Uris still work for that use case though: simply have your command open the custom document

Okay, I made a new issue: https://github.com/microsoft/vscode/issues/152716
Thank you
